tendermint/src/lite_impl/signed_header.rs

```diff
@@ -72,87 +83,43 @@ impl lite::Commit for block::signed_header::SignedHeader {
 fn non_absent_votes(commit: &block::Commit) -> Vec<vote::Vote> {
     let mut votes: Vec<vote::Vote> = Default::default();
     for (i, commit_sig) in commit.signatures.iter().enumerate() {
-        if commit_sig.is_absent() {
-            continue;
-        }
-
-        if let Some(val_addr) = commit_sig.validator_address {
-            if let Some(sig) = commit_sig.signature.clone() {
-                let vote = vote::Vote {
-                    vote_type: vote::Type::Precommit,
-                    height: commit.height,
-                    round: commit.round,
-                    block_id: Option::from(commit.block_id.clone()),
-                    timestamp: commit_sig.timestamp,
-                    validator_address: val_addr,
-                    validator_index: u64::try_from(i)
-                        .expect("usize to u64 conversion failed for validator index"),
-                    signature: sig,
-                };
-                votes.push(vote);
+        let extracted_validator_address;
+        let extracted_timestamp;
+        let extracted_signature;
+        match commit_sig {
+            CommitSig::BlockIDFlagAbsent { .. } => continue,
+            CommitSig::BlockIDFlagCommit {
+                validator_address,
+                timestamp,
+                signature,
+            } => {
+                extracted_validator_address = validator_address;
+                extracted_timestamp = timestamp;
+                extracted_signature = signature;
             }
-        }
-    }
-    votes
-}
-
-// TODO: consider moving this into commit_sig.rs instead and making it pub
-impl block::commit_sig::CommitSig {
-    fn validate(&self, vals: &Set) -> Result<(), Error> {
-        match self.block_id_flag {
-            BlockIDFlag::BlockIDFlagAbsent => {
-                if self.validator_address.is_some() {
-                    fail!(
-                        Kind::ImplementationSpecific,
-                        "validator address is present for absent CommitSig {:#?}",
-                        self
-                    );
-                }
-                if self.signature.is_some() {
-                    fail!(
-                        Kind::ImplementationSpecific,
-                        "signature is present for absent CommitSig {:#?}",
-                        self
-                    );
-                }
-                // TODO: deal with Time
-                // see https://github.com/informalsystems/tendermint-rs/pull/196#discussion_r401027989
```

Comment:
There was also a reminder about the time peculiarities :-)

Reply:
Oh sh*t, I completely missed that. :( Thanks for the thorough review!

Should I add the fix on this PR?

The fix as I see is accepting Nil commits with timestamp and checkinf if the timestamp is 0. If yes, all's well (but don't put the timestamp in the enum), if no, raise an error.

Reply:
Opened issue #259 about this.